File: camera_helper/src/org/lineageos/camerahelper/CameraMotorService.java
/*
 * Copyright (C) 2019 The LineageOS Project
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.camerahelper;

import android.annotation.NonNull;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.camera2.CameraManager;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.SystemClock;
import android.util.Log;

public class CameraMotorService extends Service implements Handler.Callback {
    private static final boolean DEBUG = true;
    private static final String TAG = "CameraMotorService";

    public static final int CAMERA_EVENT_DELAY_TIME = 100; // ms

    public static final String FRONT_CAMERA_ID = "1";

    public static final int MSG_CAMERA_CLOSED = 1000;
    public static final int MSG_CAMERA_OPEN = 1001;

    private Handler mHandler = new Handler(this);
    private CameraManager mCameraManager;
    private FallSensor mFallSensor;

    private long mClosedEvent;
    private long mOpenEvent;

    private CameraManager.AvailabilityCallback mAvailabilityCallback =
            new CameraManager.AvailabilityCallback() {
                @Override
                public void onCameraAvailable(@NonNull String cameraId) {
                    super.onCameraAvailable(cameraId);

                    if (cameraId.equals(FRONT_CAMERA_ID)) {
                        mClosedEvent = SystemClock.elapsedRealtime();
                        if (SystemClock.elapsedRealtime() - mOpenEvent < CAMERA_EVENT_DELAY_TIME
                                && mHandler.hasMessages(MSG_CAMERA_OPEN)) {
                            mHandler.removeMessages(MSG_CAMERA_OPEN);
                        }
                        mHandler.sendEmptyMessageDelayed(MSG_CAMERA_CLOSED,
                                CAMERA_EVENT_DELAY_TIME);
                    }
                }

                @Override
                public void onCameraUnavailable(@NonNull String cameraId) {
                    super.onCameraUnavailable(cameraId);

                    if (cameraId.equals(FRONT_CAMERA_ID)) {
                        mOpenEvent = SystemClock.elapsedRealtime();
                        if (SystemClock.elapsedRealtime() - mClosedEvent < CAMERA_EVENT_DELAY_TIME
                                && mHandler.hasMessages(MSG_CAMERA_CLOSED)) {
                            mHandler.removeMessages(MSG_CAMERA_CLOSED);
                        }
                        mHandler.sendEmptyMessageDelayed(MSG_CAMERA_OPEN,
                                CAMERA_EVENT_DELAY_TIME);
                    }
                }
            };

    private BroadcastReceiver mScreenStateReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            if (action.equals(Intent.ACTION_SCREEN_ON)) {
                if (DEBUG) Log.d(TAG, "Screen on, enabling fall sensor");
                mFallSensor.enable();
            } else if (action.equals(Intent.ACTION_SCREEN_OFF)) {
                if (DEBUG) Log.d(TAG, "Screen off, disabling fall sensor");
                mFallSensor.disable();
            }
        }
    };

    @Override
    public void onCreate() {
        CameraMotorController.calibrate();

        mCameraManager = getSystemService(CameraManager.class);
        mCameraManager.registerAvailabilityCallback(mAvailabilityCallback, null);

        mFallSensor = new FallSensor(this);

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(Intent.ACTION_SCREEN_OFF);
        intentFilter.addAction(Intent.ACTION_SCREEN_ON);
        registerReceiver(mScreenStateReceiver, intentFilter);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (DEBUG) Log.d(TAG, "Starting service");
        mFallSensor.enable();
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        if (DEBUG) Log.d(TAG, "Destroying service");
        mFallSensor.disable();
        unregisterReceiver(mScreenStateReceiver);
        mCameraManager.unregisterAvailabilityCallback(mAvailabilityCallback);
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public boolean handleMessage(Message msg) {
        switch (msg.what) {
            case MSG_CAMERA_CLOSED:
            case MSG_CAMERA_OPEN:
                CameraMotorController.setOpen(msg.what == MSG_CAMERA_OPEN);
                break;
        }
        return true;
    }
}

File: camera_helper/src/org/lineageos/camerahelper/CameraMotorService.kt
/*
 * Copyright (C) 2019 The LineageOS Project
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package org.lineageos.camerahelper

import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CameraManager.AvailabilityCallback
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.os.Message
import android.os.SystemClock
import android.util.Log

class CameraMotorService : Service(), Handler.Callback {
    private lateinit var cameraManager: CameraManager
    private lateinit var fallSensor: FallSensor

    private val handler = Handler(Looper.myLooper()!!, this)
    
    private val availabilityCallback: AvailabilityCallback = object : AvailabilityCallback() {
        private var closedEvent: Long = 0
        private var openEvent: Long = 0
        
        private fun handleCameraEvent(delta: Long, available: Boolean) {
            val message = if (available) MSG_CAMERA_OPEN else MSG_CAMERA_CLOSED
            if (delta < MSG_CAMERA_DELAY_MS
                && handler.hasMessages(message)
            ) {
                handler.removeMessages(message)
            }
            handler.sendEmptyMessageDelayed(
                if (available) MSG_CAMERA_CLOSED else MSG_CAMERA_OPEN,
                MSG_CAMERA_DELAY_MS
            )
        }

        override fun onCameraAvailable(cameraId: String) {
            super.onCameraAvailable(cameraId)
            if (cameraId != "1") return
            closedEvent = SystemClock.elapsedRealtime()
            handleCameraEvent(closedEvent - openEvent, true)
        }

        override fun onCameraUnavailable(cameraId: String) {
            super.onCameraUnavailable(cameraId)
            if (cameraId != "1") return
            openEvent = SystemClock.elapsedRealtime()
            handleCameraEvent(openEvent - closedEvent, false)
        }
    }

    private val screenStateReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.action) {
                Intent.ACTION_SCREEN_ON -> fallSensor.enable()
                Intent.ACTION_SCREEN_OFF -> fallSensor.disable()
            }
        }
    }

    override fun onCreate() {
        Log.d(TAG, "Creating service")
        CameraMotorController.calibrate()
        cameraManager = getSystemService(CameraManager::class.java)!!
        cameraManager.registerAvailabilityCallback(availabilityCallback, null)

        fallSensor = FallSensor(this)
        val screenStateFilter = IntentFilter()
        screenStateFilter.addAction(Intent.ACTION_SCREEN_OFF)
        screenStateFilter.addAction(Intent.ACTION_SCREEN_ON)
        registerReceiver(screenStateReceiver, screenStateFilter)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        fallSensor.enable()
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        fallSensor.disable()
        unregisterReceiver(screenStateReceiver)
        cameraManager.unregisterAvailabilityCallback(availabilityCallback)
    }

    override fun onBind(intent: Intent): IBinder? { return null }

    override fun handleMessage(msg: Message): Boolean {
        when (msg.what) {
            MSG_CAMERA_CLOSED ->
                CameraMotorController.setMotorDirection("0")
            MSG_CAMERA_OPEN ->
                CameraMotorController.setMotorDirection("1")
        }
        return true
    }

    companion object {
        private const val TAG = "CameraMotorService"
        private const val MSG_CAMERA_DELAY_MS = 100L
        private const val MSG_CAMERA_CLOSED = 1000
        private const val MSG_CAMERA_OPEN = 1001
    }
}
